patches/server/0890-Add-Alternate-Current-redstone-implementation.patch

```diff
@@ -0,0 +1,2474 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Space Walker <spacedoesrs@gmail.com>
+Date: Tue, 5 Apr 2022 01:01:13 +0200
+Subject: [PATCH] Add Alternate Current redstone implementation
+
+Author: Space Walker <spacedoesrs@gmail.com>
+
+Original license: MIT
+Original project: https://github.com/SpaceWalkerRS/alternate-current
+
+This patch adds Alternate Current's redstone implementation as an alternative to vanilla and Eigencraft's.
+Performance of (de)powering redstone dust is many times faster than vanilla, and even exceeds Eigencraft.
+Similar to Eigencraft, Alternate Current heavily changes the update order of redstone dust. This means any contraption that
+is location dependent in vanilla will either work everywhere or nowhere when using Alternate Current/Eigencraft. Beyond that
+parity issues should be rare for both implementations, though Alternate Current has not been tested as thoroughly, so I
+cannot comment on how the two compare in that aspect.
+
+Alternate Current is more invasive than Eigencraft, though some of the modifications can be removed with only a minor
+performance penalty in a small number of cases. Alternate Current needs the following modifications:
+* Level/ServerLevel: Each level has its own 'wire handler' that handles redstone dust power changes.
+* RedStoneWireBlock: Replace calls to vanilla's or Eigencraft's methods for handling power changes with calls to
+Alternate Current's wire handler.
+* (optional) Every power emitter's block class: new methods added to these classes make it easier for Alternate Current
+to check if any block is a potential power source.
+
+The use-faster-eigencraft-redstone config has been replaced with the redstone-implementation config, which can be used to
+switch between the vanilla, Eigencraft, and Alternate Current implementations.
+
+diff --git a/src/main/java/alternate/current/wire/LevelAccess.java b/src/main/java/alternate/current/wire/LevelAccess.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..3ed26ac796d9c3872ecfb572219372bcaf67e25e
+--- /dev/null
++++ b/src/main/java/alternate/current/wire/LevelAccess.java
+@@ -0,0 +1,126 @@
++package alternate.current.wire;
++
++import net.minecraft.core.BlockPos;
++import net.minecraft.core.Direction;
++import net.minecraft.server.level.ServerLevel;
++import net.minecraft.world.level.block.Block;
++import net.minecraft.world.level.block.Blocks;
++import net.minecraft.world.level.block.state.BlockState;
++import net.minecraft.world.level.chunk.ChunkAccess;
++import net.minecraft.world.level.chunk.ChunkStatus;
++import net.minecraft.world.level.chunk.LevelChunkSection;
++
++import org.bukkit.event.block.BlockRedstoneEvent;
++
++public class LevelAccess {
```

Comment:
Everything in this class but `doRedstoneEvent` and `setWireState` seem like they should just be called directly instead of the wrapper methods (either by keeping the ServerLevel instead of a LevelAccess, or with a `level()` getter) of the same exact name/a two-liner that is only used once or twice.

The remaining two could possibly also be moved into a utility class with static methods or to where they are actually used